gordon/interfaces.py

```diff
@@ -25,6 +25,9 @@ class IEventMessage(Interface):
     or cleanup in case of errors.
     """
 
+    msg_id = Attribute('Identifier for the event message instance.')
```

Comment:
Is there added value to having this defined here rather than in the `__init__` below? As I understood it, one can define required attributes like this way, or in the `__init__` method, that it's redundant to do both. I could definitely have the wrong understanding; zope docs are lacking in this area. But I think it should be one or the other, not both.

What I didn't find clear in the zope docs was if defining attributes up here was defining class attributes, rather than instance attributes. Their examples show otherwise though, so I don't think these `Attribute()`s meant strictly for class attributes. But to avoid confusion, @matt0bi and I put these in the `__init__`.

🤔 Did you find that these failed verification otherwise?

Reply:
Actually, now that I wrote it I kind of like saving it as `msg.id` better, but maybe that's just confusing given the init parameter.  I don't have any strong feelings about removing it from the init declaration. It gives the opportunity for the `id` to be determined by something that isn't passed into the init, but it's not that big of a loss for me.

Reply:
I didn't think of it that way - wish I had initially since it's a very good point. You've convinced me to go with `Attributes` like you have here, and that having it in both is fine (but may or may not be what one wants, depending on whatever the use case is). I would definitely shy away from naming anything `id` since it's a reserved keyword in Python.

I'd actually be okay with removing it (and `phase`) from `__init__`, if you do want to/isn't a lot of work, since we don't need to force a guarantee that whomever produces these messages provides upon initialization, just that those attributes are there.

Would there be harm in moving the rest of the init attributes up outside of `__init__`? I can't immediately think of any 🤔

Reply:
Phase could be removed from init, but I think I wouldn't want to require data and history_log as attributes, since we have only been using those as private variables (i.e. nothing else should access them directly at the core level, and should instead call `update_history` (and maybe later when we want it `dump_history` (or better name)) and leave the implementation of how they want to store it to the implementer.  However in the name of getting this merged in order to update gordon-gcp, I'm going to leave it as-is for now, but I will note it down for potential optimizations.